### ISSUE TITLE
Custom Markers for Workflow History/Timeline

### ISSUE THREAD

**Issue by freisenhauer** [temporalio/features#399]
**Is your feature request related to a problem? Please describe.**
I'd like to have the possibility to include workflow specific information like decisions or pivotal events inside history and/or timeline. I tried to Signal the same workflow without actually doing anything with the signal, but that's not permitted as described inside [this issue](https://github.com/temporalio/temporal/issues/682). Currently, as a workaround, we expose those information inside queries, but nothing indicates someone (trying to find an error) that there is a query that can be used and these information would likely be booleans in most cases and therefore the queries actually don't seem to be handy for this demand. 

**Describe the solution you'd like**
A way to record custom `MarkerRecorded` Events that are opaque for the server would be a suitable solution.

**Describe alternatives you've considered**
As mentioned before: Signaling (and ignoring) the own workflow which doesn't work either

**Additional context**

**Comment by Quinn-With-Two-Ns** [comment]
Moving to features since this is not a Go SDK specific issue.

Can you give some more detail on what kind of information these "markers" contain? I wonder if you have tried [UpsertMemo](https://pkg.go.dev/go.temporal.io/sdk/workflow#UpsertMemo)?

**Comment by freisenhauer** [comment]
Lets consider a workflow like 

```
[...]
if shouldDoSomething() {
  recordShouldDoSomething()
  doSomething()
}
[...]
```

then `recordShouldDoSomething()` should add information that is (easily) noticable inside Temporal UI. Ideally, directly inside history/timeline. 

I just tried to store some information inside memo, but I can't find where it's displayed in UI. Through network console i can see that my test memo information are stored and returned by API but it's not displayed anywhere. Only event in history i've noticed is `WorkflowPropertiesModified` but it's payload is `undefined` and it's not showing inside timeline.

**Comment by freisenhauer** [comment]
Just to be sure: The `MarkerRecorded` i've mentioned is the one that is documented in [Temporal Events reference](https://docs.temporal.io/references/events#markerrecorded) and hence should be already available (thats why i opened the issue inside go sdk).

**Comment by cretz** [comment]
This seems like a reasonable ask to me. This is like #141 but a whole new event instead of just metadata on an existing command. We have resisted log shipping of course, but someone wanting to add a "user-defined event" I think seems reasonable. I have ideas about implementation, but yes it should probably be in history as a marker.

**Comment by Quinn-With-Two-Ns** [comment]
Yeah if we want timeline support memo is no appropriate it would need to be some custom marker that the UI is also aware of.

>Just to be sure: The MarkerRecorded i've mentioned is the one that is documented in [Temporal Events reference](https://docs.temporal.io/references/events#markerrecorded) and hence should be already available (thats why i opened the issue inside go sdk).

No SDK provides an API to directly write custom markers. They are meant to be used internally by the SDKs to implement certain features  like SideEffects and Local Activities and this may be another one.

**Comment by freisenhauer** [comment]
Okay, understandable. 

Sounds like in general the idea meets with approval, but needs amendments through to a greater part of the stack, right ?

> They are meant to be used internally by the SDKs to implement certain features like SideEffects and Local Activities and this may be another one.

This leaves me hopeful, we at least don't have to do something to the temporal server.

Without knowing your exact contribution guidelines or the code I could at least _offer_ support in this concern (to whatever extend) 😆

**Comment by freisenhauer** [comment]
Could you explain why you think this is unsafer than e.g. introducing a basically noop Activity that gets a payload with description and is only used to be viewed in history ? What is stopping one from doing this and why should this decision be made by the product and not the developer using it ? 

Doing things in (endless) loop is also something that can lead to history growth and yet it is even documented, how to handle this in practice. This does not really seem to be consistent with previous decisions and/or documented use cases.

**Comment by mfateev** [comment]
We don't want to encourage behavior that would lead to performance issues. Most users don't think about the cost of these features and then get surprised that recovery takes longer. 

At the same time we absolutely need to solve the use case of waiting for workflow reaching a certain state. The current workaround is to use an update for this.

**Comment by freisenhauer** [comment]
Okay, I can understand that this does not have highest prio at the moment. Could you explain what you mean by "use an update for this" ?

**Comment by mfateev** [comment]
i meant waiting for workflow to reach a specific state.

**Comment by freisenhauer** [comment]
ah okay, I understand.
